blogpost: Introducing the migrator service (#5070)
sourcegraph/about: blogposts/2022/introducting-migrator-service.md

---
title: 'Introducing the migrator service'
author: Eric Fritz
authorUrl: https://eric-fritz.com
publishDate: 2022-02-15T12:00-00:00
tags: [blog]
slug: introducing-migrator-service
heroImage: https://sourcegraphstatic.com/blog/introducing-migrator-service/introducing-migrator-service.jpg
socialImage: https://sourcegraphstatic.com/blog/introducing-migrator-service/introducing-migrator-service.jpg
published: true
description: "We have identified and solved a huge source of pain during upgrades for Sourcegraph administrators. Here's how."
---

> We have recently identified a root cause of a very frequent and insidious pain related to database migrations during Sourcegraph version upgrades. Many users handling the installation and upgrade of their Sourcegraph instance can likely attest to seeing an error similar to the following mid-way through an upgrade from one minor release to the next.

Author: Oh, boy. If there is some loosely gathered stat I could use like "dirty databases occurred for X upgrades during Y in customer instances" that would be super great.

Or if there's anything that stands out to mention here as a particular heinous example (data loss or something similar would be catastrophically delicious).

> We have recently identified a root cause of a very frequent and insidious pain related to database migrations during Sourcegraph version upgrades. Many users handling the installation and upgrade of their Sourcegraph instance can likely attest to seeing an error similar to the following mid-way through an upgrade from one minor release to the next.

Reviewer: https://github.com/sourcegraph/customer/issues/543 -- this one comes to mind, but, I know there are more, will inquire with team for more issues. I don't think our stats track issues by anything more specific than department, and this would fall under delivery issues in our current tracking

---
title: 'Introducing the migrator service'
author: Eric Fritz
authorUrl: https://eric-fritz.com
publishDate: 2022-02-15T12:00-00:00
tags: [blog]
slug: introducing-migrator-service
heroImage: https://sourcegraphstatic.com/blog/introducing-migrator-service/introducing-migrator-service.jpg

> socialImage: https://sourcegraphstatic.com/blog/introducing-migrator-service/introducing-migrator-service.jpg

Author: - [ ] Need her+social images.

Reviewer 2: @efritz I have a request out for images but we'll have to add them after the fact. It will be at least a few days, if not a week, unless maybe you have some temporary idea? Like a screenshot or diagram or something?

Author: I don't have anything prepared here, sorry 😿

Reviewer 2: No worries, didn't expect you to! I sent over some concepts to our graphic team so we'll be able to add something in post in like a week.

Author: Do we need to put a placeholder here? I'm not sure it can be blank.

Reviewer 2: Oh, I'm not actually sure! Looking back it does look like even old blog posts pre-graphics help had something

Reviewer 2: Not exciting but we could just do the logo https://about.sourcegraph.com/sourcegraph-mark.png

```sql
[A] sourcegraph@sourcegraph ❯ CREATE INDEX CONCURRENTLY test ON repo(name, deleted_at);
ERROR:  40P01: deadlock detected
DETAIL:  Process 2149 waits for ShareLock on virtual transaction 3/13053932; blocked by process 2471.
Process 2471 waits for ExclusiveLock on advisory lock [16384,0,0,2]; blocked by process 2149.
HINT:  See server log for query details.
LOCATION:  DeadLockReport, deadlock.c:1153
Time: 1107.901 ms (00:01.108)
```

This error is a bit dense, but states that **Process A** and **Process B** are holding locks the other one wants. The concurrent index creation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but is blocked on **Process A**'s ownership of that lock.

Additionally, this leaves the database with a _partially created_ index, which exists in the system catalog but is available neither for reads nor writes. To resolve this database state previously, an administrator must drop and recreate the index manually. This is an easy step to forget, and one that can have very severe performance penalties. In newer versions of the migrator, invalid indexes are automatically detected and repaired.

> TODO - need more content here

Author: - [ ] Tie-in back to original error and add a more exciting introduction to the second half of the solution

The migrators now gracefully deal with concurrent index creation by dropping the advisory lock and instead coordinating with concurrent migration attempts within the Postgres catalog itself. Of particular interest is the `pg_stat_progress_create_index` system view, which [reports index creation progress](https://www.postgresql.org/docs/12/progress-reporting.html).

## Conclusions

> I don't know. Here's some links.

Author: Conclusion not yet written.

With a new theory in hand, we can finally attack the problem in a focused manner.

We changed the migration process to gracefully deals with concurrent index creation by dropping the advisory lock and instead coordinating with concurrent migration attempts within the Postgres catalog itself. Of particular interest is the `pg_stat_progress_create_index` system view, which [reports index creation progress](https://www.postgresql.org/docs/12/progress-reporting.html). The migrator is also intelligence enough to detect invalid indexes are repair them rather than flagging a human to hold its hand.

After another round of testing, migrations finally behave according to intuition.

## Conclusion

We've long since outgrown the way in which we run database migrations. We've known for a while that there _was_ a problem, and felt the pain along with our users for an embarrassingly long time. After digging a bit deeper, better understand the problem space, and breaking a bit of our ossified code and deployment process along the way, I hope we're delivering a quality of life improvement to Sourcegraphers and our users.

> If your next upgrade does **not** crash, please tweet me at [@ericfritz](https://twitter.com/ericfritz).

Author: Is this ... ok to do? :laff:

Reviewer: I approve

### Related resources

We've updated several documentation pages which are worth a look for existing site administrators:

> - [How to run `migrator` operations](https://docs.sourcegraph.com/admin/how-to/manual_database_migrations)

Author: @DaedalusG Any other docs you'd like to call out in-particular?

Author: re. https://github.com/sourcegraph/about/pull/5070#discussion_r808309950

Reviewer: Hmm actually that would be the main one - perhaps this doc
https://docs.sourcegraph.com/admin/how-to/rollback_database

I mainly made this comment after running into a few folks who were confused about the term migratory and assumed it was related to database migrations between databases

[https://sourcegraph.slack.com/archives/C02BJ8T258D/p1644968878319539](https://sourcegraph.slack.com/archives/C02BJ8T258D/p1644968878319539)
[https://sourcegraph.slack.com/archives/C01JR51JR5J/p1645032270381909](https://sourcegraph.slack.com/archives/C01JR51JR5J/p1645032270381909)

- [How to troubleshoot a dirty database](https://docs.sourcegraph.com/admin/how-to/dirty_database) (replacing the [legacy documentation](https://docs.sourcegraph.com/admin/how-to/dirty_database_pre_3_37))

> We are open source! Check out some of the higher-scope code changes that made this change possible:

Author: Is there a better term or approved copy to use for things like this?

Reviewer 2: I'm not sure! Will check

Author: Didn't want to attribute open-core to a non-(truly)-open-core tech. Seems like being precise with what we identify ourselves as here is necessary to prevent some bad opinions from forming when we post to places like HN.

Reviewer 2: Agreed, I think precision is key. Kacie said we're not really open source, so I'm wondering if it's better to say something like "We support open source!" We could also omit that statement and just tell readers they can check out the code changes.

Author: ```suggestion
Check out some of the higher-scope code changes that made this change possible:
```

---
title: 'Introducing the migrator service'
author: Eric Fritz
authorUrl: https://eric-fritz.com
publishDate: 2022-02-15T12:00-00:00
tags: [blog]
slug: introducing-migrator-service
heroImage: https://sourcegraphstatic.com/blog/introducing-migrator-service/introducing-migrator-service.jpg
socialImage: https://sourcegraphstatic.com/blog/introducing-migrator-service/introducing-migrator-service.jpg
published: true
description: "We have identified and solved a huge source of pain during upgrades for Sourcegraph administrators. Here's how."
---

> We have recently identified the root cause of a very frequent and insidious pain related to database migrations during Sourcegraph version upgrades. Many users handling the installation and upgrade of their Sourcegraph instance can likely attest to seeing an error similar to the following mid-way through an upgrade from one minor release to the next.

Reviewer 2: ```suggestion
We have recently identified the root cause of a frequent, insidious pain-point related to database migrations during Sourcegraph version upgrades. Many users handling the installation and upgrade of their Sourcegraph instance can likely attest to seeing an error similar to the following mid-way through an upgrade from one minor release to the next.
```

```log
ERROR: Failed to migrate the DB. Please contact support@sourcegraph.com for further assistance: Dirty database version 1528395797. Fix and force version.
```

> This error message is the bane of any engineer in our customer support org, any engineer in our #opsgenie Slack channel, and a growing number of site administrators.

Reviewer 2: ```suggestion
This error message is the bane of a growing number of site administrators as well as any engineer in our customer support org or any engineer in our #opsgenie Slack channel.
```

> As of Sourcegraph 3.37, we no longer spuriously mark the database as dirty during common upgrade scenarios. We have decoupled database migrations from the application startup sequence to better control database schema upgrade conditions.

Reviewer 2: ```suggestion
As of Sourcegraph 3.37, we no longer spuriously mark the database as dirty during common upgrade scenarios. We have decoupled database migrations from the application start up sequence to better control database schema upgrade conditions.
```

> This release brings:

Reviewer 2: ```suggestion
This release adds:
```

> - [New tools](https://docs.sourcegraph.com/admin/how-to/manual_database_migrations) that allow site administrators to run database migrations independently from an upgrade. Large instance upgrades can be upgraded ahead-of-time of an upgrade to reduce the possibility of downtime due to migrations taking longer than the healthcheck timeout.

Reviewer 2: ```suggestion
- [New tools](https://docs.sourcegraph.com/admin/how-to/manual_database_migrations) that allow site administrators to run database migrations independently from an upgrade. You can now make large instance upgrades before a general upgrade to reduce the possibility of downtime due to migrations taking longer than the healthcheck timeout.
```
Is "general" the right word here? Let me know if I've missed your meaning.

Author: ```suggestion
- [New tools](https://docs.sourcegraph.com/admin/how-to/manual_database_migrations) that allow site administrators to run database migrations independently from an upgrade. Large database instances can be upgraded ahead-of-time of a general application upgrade to reduce the possibility of downtime due to migrations taking longer than the healthcheck timeout.
```

Author: @nickmyyz This better?

> - Better error messages and [a better recovery process](https://docs.sourcegraph.com/admin/how-to/dirty_database) in the (now appropriately rare) case of a real database migration failure.

Reviewer 2: ```suggestion
- Better error messages and [a better recovery process](https://docs.sourcegraph.com/admin/how-to/dirty_database) in the (now rare) case of a real database migration failure.
```

- A general pruning of [now-irrelevant troubleshooting documentation](https://docs.sourcegraph.com/admin/how-to/dirty_database_pre_3_37).

We've tried to make this transition as seamless as possible, and there should be no manual steps required to proceed to Sourcegraph 3.37.0 and beyond. If we've done everything right, you'll notice no differences.

But, if you'd like to see how the sausage is debugged, read on.

## The issue

> Upgrades of Sourcegraph instances, referring to both on-premise deployment and our [Cloud](https://sourcegraph.com/search) environment, would often fail with a dreaded "dirty database" error. This class of error leaves the Sourcegraph instance in a broken state that requires manual intervention to resolve.

Reviewer 2: ```suggestion
Upgrades of Sourcegraph instances, including on-premise deployments and our [Cloud](https://sourcegraph.com/search) environment, would often fail with a dreaded "dirty database" error. This class of error left the Sourcegraph instance in a broken state that required manual intervention to resolve.
```

> On the customer side, the resolution is very inelegant. A new container will come online and attempt to migrate the database. When this attempt fails, the database is marked as dirty and the container will restart. New instances will see the dirty flag set and also restart. Site-administrators will notice rapidly restarting containers (that, or a wonky Sourcegraph experience) and eventually might find a log message asking them to contact support for assistance.

Reviewer 2: ```suggestion
On the customer side, the resolution was inelegant. A new container would come online and attempt to migrate the database. When this attempt failed, the database was marked as dirty and the container would restart. New instances would see the dirty flag set and also restart. Site-administrators would notice rapidly restarting containers (that, or a wonky Sourcegraph experience) and would sometimes find a log message asking them to contact support for assistance.
```

> In our Cloud environment, Sourcegraph engineers are subjected to the same manual process.

Reviewer 2: ```suggestion
In our Cloud environment, Sourcegraph engineers endured the same manual process.
```

Requiring site administrators to directly modify database contents is not a distribution practice we should encourage. The frequency of these errors and the rising burden put on users and application/product engineers made this a pressing issue to resolve.

## Identifying (a) contributing factor

In our original effort to erradicate this class of error, we incorrectly attributed the Docker container health check period as the primary cause. Our intuition was built from two observations:

> 1. We never see this issue in local development which are usually operate over relatively small sets of data.

Reviewer 2: ```suggestion
1. We never saw this issue in local development, which usually operates over relatively small sets of data.
```

> 2. We see this issue frequently in our Cloud environment, but exclusively on objects over a particular (but undetermined) size, such as our repository or code intelligence data tables.

Reviewer 2: ```suggestion
2. We saw this issue frequently in our Cloud environment, but exclusively on objects over a particular (but undetermined) size, such as our repository or code intelligence data tables.
```

We believed that the issue occurred if and only if the set of migrations running on application startup could not finish within the parent container's health check timeout. If the migration took longer to perform than the configured timeout, Kubernetes would declare the pod dead and restart it. Killing a pod _during_ an active migration marks the database as dirty until the site administrator manually resolves the issue.

And this is exactly the behavior we see.

> Our original solution, simply "decoupling" the database from application upgrades, targeted the Kubernetes health check as the primary problem. By performing the migrations separately from the application deployment, the migration process no longer has to complete within a configured healthcheck timeout of a downstream application. By having a separate component performing the migrations, and by moving away from our dependency on a [third-party](https://github.com/golang-migrate/migrate) migration library, we can better inspect the health of the migrator and the progress of its migration attempts.

Reviewer 2: ```suggestion
Our original solution, simply "decoupling" the database from application upgrades, targeted the Kubernetes health check as the primary problem. By performing the migrations separately from the application deployment, the migration process no longer has to complete within a configured healthcheck timeout of a downstream application. By having a separate component perform the migrations, and by moving away from our dependency on a [third-party](https://github.com/golang-migrate/migrate) migration library, we can better inspect the health of the migrator and the progress of its migration attempts.
```

> The first draft of the `migrator` service kept all other behavior the same. We explicitly kept the new migration behavior as close to the previous migration behavior as possible - just running in a different container - to avoid turning too many knobs at once. Part of the behavior we explicitly did _not_ change was how we handle concurrent migration attempts.

Reviewer 2: ```suggestion
The first draft of the `migrator` service kept all other behavior the same. We explicitly kept the new migration behavior as close to the previous migration behavior as possible–just running in a different container–to avoid turning too many knobs at once. Part of the behavior we explicitly did _not_ change was how we handle concurrent migration attempts.
```

> Each migrator attempts to take an advisory lock over a shared key representing the target schema. If the lock is acquired, the current database schema version is read. If the database is marked as dirty, the migrator panics. If the database is up-to-date, the migrator can exit. Otherwise, all unappalied migration files are applied and the lock is released.

Reviewer 3: ```suggestion
Each migrator attempts to take an advisory lock over a shared key representing the target schema. If the lock is acquired, the current database schema version is read. If the database is marked as dirty, the migrator panics. If the database is up-to-date, the migrator can exit. Otherwise, all unapplied migration files are applied and the lock is released.
```

After some initial testing, we discovered that the frequency of "dirty database" behaviors did not actually decrease. The solution did not address the symptoms, and we needed to do more.

## A deeper contributing factor

> Over time, our engineering organization formed some tribal knowledge regarding ways to avoid "taking down production" with poorly defined migrations. Of particular hesitation was creating indexes on tables that can have a large number of rows. Normal index creation operations take an exclusive write lock on the table, creating the new index from the existing table data synchronously. During this time no inserts, updates, or deletes can be made to the table. You basically incur write downtime (along with a blip of failed requests in your metrics dashboards and alerting systems, and a slice of failed jobs that failed in the background) for the table, which can take tens of minutes for large tables.

Reviewer 2: ```suggestion
Over time, our engineering organization formed some tribal knowledge regarding ways to avoid "taking down production" with poorly defined migrations. Creating indexes on tables that can have a large number of rows caused many to hesitate. Normal index creation operations take an exclusive write lock on the table, creating the new index from the existing table data synchronously. During this time no inserts, updates, or deletes can be made to the table. You basically incur write downtime (along with a blip of failed requests in your metrics dashboards and alerting systems, and a slice of failed jobs that failed in the background) for the table, which can take tens of minutes for large tables.
```

For our use case, `CREATE INDEX CONCURRENTLY` was received as a hero.

Using concurrent index creation reduced downtime incidents in our Cloud environment related to heavy database contention triggered by continuous deployment.

> Using concurrent index creation also introduced a footgun and plenty of friendly fire that we didn't recognize until recently. [Concurrent index creation](https://www.postgresql.org/docs/12/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY) is performed in three distinct phases to ensure that a full table lock is never required, thus readers nor writers are not blocked for long stretches of time.

Reviewer 2: ```suggestion
Using concurrent index creation also introduced a footgun and plenty of friendly fire that we didn't recognize until recently. [Concurrent index creation](https://www.postgresql.org/docs/12/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY) is performed in three distinct phases to ensure that a full table lock is never required. That way, neither readers nor writers are not blocked for long stretches of time.
```

- **Phase 1**: Create new catalog structures for the index, but do not begin to populate them yet. Transactions that started before the completion of this phase will only see (and therefore only write to) the old catalog structures.
- **Phase 2**: Wait for all transactions starting before the end of **Phase 1** to finish. Then, perform a full-table scan and write the content of each row to the new index structure. Open the new index structure up to writes.
- **Phase 3**: Wait for all transactions starting before the end of **Phase 2** to finish. Then, perform another full-table scan and write the content of each row inserted during **Phase 2** to the new index structure. The old index structure can now be closed to writes.

> There's a particular detail here that deserves special attention in hindsight: concurrent index creation operations block until **all transactions started before them** have either committed or rolled back. Mixed with long-held advisory locks, we encounter this rather nasty situation.

Reviewer 2: ```suggestion
There's a particular detail here that deserves special attention in hindsight: concurrent index creation operations block until **all transactions started before them** have either committed or rolled back. Mixed with long-held advisory locks, we encounter the rather nasty situation below.
```

Observe! In `psql` **Process A**, we acquire an advisory lock:

```sql
[A] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
 pg_advisory_lock
------------------

(1 row)

Time: 20.616 ms
```

In `psql` **Process B**, we attempt to acquire an advisory lock:

```sql
[B] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);

```

> As this lock is already held by another process, this command blocks indefinitely - or until the user gets bored and aborts their session.

Reviewer 2: ```suggestion
Because this lock is already held by another process, this command blocks indefinitely–or until the user gets bored and aborts their session.
```

Back in `psql` **Process A**, we attempt to create an index concurrently on _literally any table_:

```sql
[A] sourcegraph@sourcegraph ❯ CREATE INDEX CONCURRENTLY test ON repo(name, deleted_at);
ERROR:  40P01: deadlock detected
DETAIL:  Process 2149 waits for ShareLock on virtual transaction 3/13053932; blocked by process 2471.
Process 2471 waits for ExclusiveLock on advisory lock [16384,0,0,2]; blocked by process 2149.
HINT:  See server log for query details.
LOCATION:  DeadLockReport, deadlock.c:1153
Time: 1107.901 ms (00:01.108)
```

> After a second or so, the Postgres deadlock detector kicks in, notices two processes vying for each other's locks, and cancels the deadlocked index creation operation. This error is a bit dense, but states that the index creation operation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but the advisory lock acquisition is blocked on **Process A**'s ownership of that lock. Concurrent index creation failures are also a paritcularly nasty failure mode, as they leave the database with a _partially created_ index that exists in the system catalog but is accessible neither for reads nor writes. To resolve this database state, the invalid index has to be dropped and re-created. This is an easy step to forget, if required to be done manually, and forgetting the step can have very severe performance penalties if not detected.

Reviewer 2: ```suggestion
After a second or so, the Postgres deadlock detector kicks in, notices two processes vying for each other's locks, and cancels the deadlocked index creation operation. This error is a bit dense, but states that the index creation operation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but the advisory lock acquisition is blocked on **Process A**'s ownership of that lock. 

Concurrent index creation failures are also a particularly nasty failure mode, because they leave the database with a _partially created_ index that exists in the system catalog but is accessible to reads or writes. To resolve this database state, the invalid index has to be dropped and re-created. This is an easy step to forget you need to do it manually, and forgetting the step can have severe performance penalties if you don't detect it.
```

With a new theory in hand, we can finally attack the problem in a focused manner.

We changed the migration process to gracefully deals with concurrent index creation by dropping the advisory lock and instead coordinating with concurrent migration attempts within the Postgres catalog itself. Of particular interest is the `pg_stat_progress_create_index` system view, which [reports index creation progress](https://www.postgresql.org/docs/12/progress-reporting.html). The migrator is also intelligence enough to detect invalid indexes are repair them rather than flagging a human to hold its hand.

After another round of testing, migrations finally behave according to intuition.

## Conclusion

> We've long since outgrown the way in which we run database migrations. We've known for a while that there _was_ a problem, and felt the pain along with our users for an embarrassingly long time. After digging a bit deeper, better understand the problem space, and breaking a bit of our ossified code and deployment process along the way, I hope we're delivering a quality of life improvement to Sourcegraphers and our users.

Reviewer 2: ```suggestion
We've long since outgrown the way in which we run database migrations. We've known for a while that there _was_ a problem, and felt the pain along with our users for an embarrassingly long time. After digging a bit deeper, better understanding the problem space, and breaking a bit of our ossified code and deployment process along the way, we can now deliver a significant quality of life improvement to our users and Sourcegraphers alike.
```

---

> title: 'Introducing the migrator service'

Reviewer 4: Yeah, I'm thinking of some. HN drops "introducing" so we might need to be a little bit more descriptive. cc @rebeccadee

> title: 'Introducing the migrator service'

Reviewer 2: Spitballing, how about:
- How we fixed database migrations during version upgrades
- How breaking our ossified code and deployment process helped us fix migration issues
- Broken database migrations: How we finally fixed an embarrassing problem
- How (and why) we decoupled database migrations from the application start up sequence

> title: 'Introducing the migrator service'

Author: I like the truth to _Broken database migrations: How we finally fixed an embarrassing problem_

> title: 'Introducing the migrator service'

Reviewer 4: @efritz I love that! Puts mine to shame:

* Making database migrations easier for our users

> title: 'Introducing the migrator service'

Author: ```suggestion
title: 'Broken database migrations: How we finally fixed an embarrassing problem'
```

## The issue

Upgrades of Sourcegraph instances, including on-premise deployments and our [Cloud](https://sourcegraph.com/search) environment, would often fail with a dreaded "dirty database" error. This class of error left the Sourcegraph instance in a broken state that required manual intervention to resolve.

On the customer side, the resolution was inelegant. A new container would come online and attempt to migrate the database. When this attempt failed, the database was marked as dirty and the container would restart. New instances would see the dirty flag set and also restart. Site-administrators would notice rapidly restarting containers (that, or a wonky Sourcegraph experience) and would sometimes find a log message asking them to contact support for assistance.

In our Cloud environment, Sourcegraph engineers endured the same manual process.

Requiring site administrators to directly modify database contents is not a distribution practice we should encourage. The frequency of these errors and the rising burden put on users and application/product engineers made this a pressing issue to resolve.

## Identifying (a) contributing factor

> In our original effort to erradicate this class of error, we incorrectly attributed the Docker container health check period as the primary cause. Our intuition was built from two observations:

Author: I would say uses of `healthcheck(s)` -> `health check(s)`.

> In our original effort to erradicate this class of error, we incorrectly attributed the Docker container health check period as the primary cause. Our intuition was built from two observations:

Author: Updated in two places.

Observe! In `psql` **Process A**, we acquire an advisory lock:

```sql
[A] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);
 pg_advisory_lock
------------------

(1 row)

Time: 20.616 ms
```

In `psql` **Process B**, we attempt to acquire an advisory lock:

```sql
[B] sourcegraph@sourcegraph ❯ SELECT pg_advisory_lock(0, 0);

```

Because this lock is already held by another process, this command blocks indefinitely–or until the user gets bored and aborts their session.

Back in `psql` **Process A**, we attempt to create an index concurrently on _literally any table_:

```sql
[A] sourcegraph@sourcegraph ❯ CREATE INDEX CONCURRENTLY test ON repo(name, deleted_at);
ERROR:  40P01: deadlock detected
DETAIL:  Process 2149 waits for ShareLock on virtual transaction 3/13053932; blocked by process 2471.
Process 2471 waits for ExclusiveLock on advisory lock [16384,0,0,2]; blocked by process 2149.
HINT:  See server log for query details.
LOCATION:  DeadLockReport, deadlock.c:1153
Time: 1107.901 ms (00:01.108)
```

> After a second or so, the Postgres deadlock detector kicks in, notices two processes vying for each other's locks, and cancels the deadlocked index creation operation. This error is a bit dense, but states that the index creation operation is waiting for **Process B**'s attempt to acquire the advisory lock to finish, but the advisory lock acquisition is blocked on **Process A**'s ownership of that lock.

Author: Should we retroactively update the uses in https://about.sourcegraph.com/blog/postgres-version-update/?
